initialize/user_data.go

```diff
@@ -15,40 +15,410 @@
+			contents:     payload,
+			fileName:     "userdata.txt",
+		})
 	}
+
+	return parts, nil
+}
+
+type UserDataPart struct {
+	userDataType UserDataType
+	contents     string
+	fileName     string
+
+	cloudConfig *config.CloudConfig
+	script      *config.Script
+}
+
+func (udp *UserDataPart) PartType() UserDataType {
+	return udp.userDataType
+}
+
+func (udp *UserDataPart) PartName() string {
+	return udp.fileName
+}
+
+func (udp *UserDataPart) IsCloudConfig() bool {
+	return udp.userDataType == CloudConfigType
+}
+
+func (udp *UserDataPart) IsScript() bool {
+	return udp.userDataType == ScriptType
+}
+
+func (udp *UserDataPart) IsIgnition() bool {
+	return udp.userDataType == IgnitionType
+}
+
+func (udp *UserDataPart) IsUnknown() bool {
+	return udp.userDataType == UnknownType
+}
+
+func (udp *UserDataPart) runScript(env *Environment) error {
+	if env == nil {
+		return fmt.Errorf("environment is nil")
+	}
+
+	err := PrepWorkspace(env.Workspace())
+	if err != nil {
+		log.Printf("Failed preparing workspace: %v\n", err)
+		return err
+	}
+	path, err := PersistScriptInWorkspace(*udp.script, env.Workspace())
+	if err == nil {
+		var name string
+		name, err = system.ExecuteScript(path)
+		PersistUnitNameInWorkspace(name, env.Workspace())
+	}
+	return err
+}
+
+func (udp *UserDataPart) runCloudConfig(env *Environment) error {
+	if err := Apply(*udp.cloudConfig, env); err != nil {
+		return fmt.Errorf("error applying cloud-config: %w", err)
+	}
+	return nil
+}
+
+func (udp *UserDataPart) RunPart(env *Environment) error {
+	switch udp.userDataType {
+	case ScriptType:
+		return udp.runScript(env)
+	case CloudConfigType:
+		return udp.runCloudConfig(env)
+	default:
+		log.Printf("ignoring part of type %s", udp.userDataType)
+	}
+	return nil
+}
+
+type UserData struct {
+	Parts []UserDataPart
+
+	env *Environment
+}
+
+func (ud *UserData) FindHostname() string {
+	for _, part := range ud.Parts {
+		if part.cloudConfig != nil && part.cloudConfig.Hostname != "" {
+			return part.cloudConfig.Hostname
+		}
+	}
+	return ""
+}
+
+func (ud *UserData) FindSSHKeys(additionalKeys []string) []string {
+	keys := []string{}
+	hasKey := func(key string) bool {
+		for _, k := range keys {
+			if k == key {
+				return true
+			}
+		}
+		return false
+	}
+
+	for _, part := range ud.Parts {
+		if part.cloudConfig != nil {
+			keys = append(keys, part.cloudConfig.SSHAuthorizedKeys...)
+
+			if part.cloudConfig.Users != nil {
+				for _, user := range part.cloudConfig.Users {
+					if user.Name != "core" {
+						continue
+					}
+
+					// The "core" user is the default user on coreos systems. Append these keys to the list.
+					for _, key := range user.SSHAuthorizedKeys {
+						if !hasKey(key) {
+							keys = append(keys, key)
+						}
+					}
+				}
+			}
+		}
+	}
+	return append(keys, additionalKeys...)
+}
+
+type headerGetter interface {
+	Get(key string) string
+}
+
+func parseMimeHeader[T headerGetter](header T) (headerInfo, error) {
```

Comment:
Can't this be `func parseMimeHeader(header headerGetter) (headerInfo, error)`?

Reply:
I think it can. Should work the same way. Will update the PR soon.